File: kotlin_codeql/src/main/java/org/example/kotlin_codeql/Security/CWE/CWE-190/ArithmeticUncontrolled.kt
// {fact rule=arithmetic-overflow@v1.0 defects=0}
import java.security.SecureRandom

internal object ArithmeticUncontrolled {
    @JvmStatic
    fun main(args: Array<String>) {
        run {
            val data = SecureRandom().nextInt()

            // BAD: may overflow if data is large
            val scaled = data * 10

            // ...

            // GOOD: use a guard to ensure no overflows occur
            val scaled2: Int
            scaled2 = if (data < Int.MAX_VALUE / 10) data * 10 else Int.MAX_VALUE
        }
    }
}
// {/fact}

File: cpp_codeql/Security/CWE/CWE-190/ArithmeticUncontrolled.cpp
#include <iostream>
#include <climits>


namespace {
    int main(int argc, char** argv) {
        int i = rand();
        // {fact rule=integer-overflow@v1.0 defects=1}
        // BAD: potential overflow
        int j = i + 1000;
        // {/fact}

        // ...

        int n = rand();
        int k;
        // {fact rule=integer-overflow@v1.0 defects=0}
        // GOOD: use a guard to prevent overflow
        if (n < INT_MAX-1000)
            k = n + 1000;
        else
            k = INT_MAX;
        // {/fact}
    }

}
